pkg/daemon/update.go

```diff
@@ -1189,16 +1186,7 @@ func (dn *CoreOSDaemon) switchKernel(oldConfig, newConfig *mcfgv1.MachineConfig)
 		dn.logSystem("Switching to kernelType=%s, invoking rpm-ostree %+q", newConfig.Spec.KernelType, args)
 		return runRpmOstree(args...)
 	}
-
-	if canonicalizeKernelType(oldConfig.Spec.KernelType) == ctrlcommon.KernelTypeRealtime && canonicalizeKernelType(newConfig.Spec.KernelType) == ctrlcommon.KernelTypeRealtime {
```

Comment:
This logic was getting used during an OSUpdate when nodes in the pool were already on RT kernel.

Reply:
To avoid confusion, we may want to preserve this logging message. Right now, we will see log something like "Switching to kernelType=Realtime, invoking rpm-ostree ... " during OS update as well.

Reply:
Ah yes you are right that the `rebase` option wouldn't upgrade before (we should have an option for that).  But this new logic will effectively do an upgrade by installing the latest again.

Hmm, current code is logging actually *three* times since we always log the args from `runRpmOstree` anyways.  So this should trim it to two, and just changes the log message to distinguish between switching and re-applying.  Does the below do what you want?  (Doing this here to avoid a very costly re-run of CI just for logging changes)

```
diff --git a/pkg/daemon/update.go b/pkg/daemon/update.go
index 2a05209bc..3a6c58518 100644
--- a/pkg/daemon/update.go
+++ b/pkg/daemon/update.go
@@ -1151,13 +1151,14 @@ func (dn *CoreOSDaemon) applyExtensions(oldConfig, newConfig *mcfgv1.MachineConf
 
 // switchKernel updates kernel on host with the kernelType specified in MachineConfig.
 // Right now it supports default (traditional) and realtime kernel
-func (dn *CoreOSDaemon) switchKernel(newConfig *mcfgv1.MachineConfig) error {
+func (dn *CoreOSDaemon) switchKernel(oldConfig, newConfig *mcfgv1.MachineConfig) error {
 	// We support Kernel update only on RHCOS and SCOS nodes
 	if !dn.os.IsEL() {
 		glog.Info("updating kernel on non-RHCOS nodes is not supported")
 		return nil
 	}
 
+	oldKtype := canonicalizeKernelType(oldConfig.Spec.KernelType)
 	newKtype := canonicalizeKernelType(newConfig.Spec.KernelType)
 
 	// In the OS update path, we removed overrides for kernel-rt.  So if the target (new) config
@@ -1173,7 +1174,11 @@ func (dn *CoreOSDaemon) switchKernel(newConfig *mcfgv1.MachineConfig) error {
 	// kernel-rt also has a split off kernel-rt-kvm subpackage because it's in a separate subscription in RHEL.
 	realtimeKernel := []string{"kernel-rt-core", "kernel-rt-modules", "kernel-rt-modules-extra", "kernel-rt-kvm"}
 
-	dn.logSystem("Initiating switch to kernel %s", newKtype)
+	if oldKtype != newKtype {
+		dn.logSystem("Initiating switch to kernel %s", newKtype)
+	} else {
+		dn.logSystem("Re-applying kernel type %s", newKtype)
+	}
 
 	if newKtype == ctrlcommon.KernelTypeRealtime {
 		// Switch to RT kernel
@@ -1183,7 +1188,6 @@ func (dn *CoreOSDaemon) switchKernel(newConfig *mcfgv1.MachineConfig) error {
 			args = append(args, "--install", pkg)
 		}
 
-		dn.logSystem("Switching to kernelType=%s, invoking rpm-ostree %+q", newConfig.Spec.KernelType, args)
 		return runRpmOstree(args...)
 	}
 	return fmt.Errorf("Unhandled kernel type %s", newKtype)
@@ -2104,7 +2108,7 @@ func (dn *CoreOSDaemon) applyLayeredOSChanges(mcDiff machineConfigDiff, oldConfi
 	}
 
 	// Switch to real time kernel
-	if err := dn.switchKernel(newConfig); err != nil {
+	if err := dn.switchKernel(oldConfig, newConfig); err != nil {
 		return err
 	}
 
@@ -2176,7 +2180,7 @@ func (dn *CoreOSDaemon) applyLegacyOSChanges(mcDiff machineConfigDiff, oldConfig
 	}
 
 	// Switch to real time kernel
-	if err := dn.switchKernel(newConfig); err != nil {
+	if err := dn.switchKernel(oldConfig, newConfig); err != nil {
 		return err
 	}

```

Reply:
yes, diff looks good. Indeed rerunning ci is expensive for small changes, so thanks for creating the diff here.

Reply:
Done